### PR TITLE
GH-41667: [C++][Parquet] Refuse writing non-nullable column that contains nulls

### PR DESCRIPTION
**Pull request by pitrou**
### Rationale for this change

A non-nullable column that contains nulls would result in an invalid Parquet file, so we'd rather raise an error when writing.

This detection is only implemented for leaf columns. Implementing it for non-leaf columns would be more involved, and also doesn't actually seem necessary.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Raising a clear error when trying to write invalid data to Parquet, instead of letting the Parquet writer silently generate an invalid file.

* GitHub Issue: #41667

### CODE REVIEW

File: cpp/src/parquet/column_writer.cc [hunk 1]
```diff
@@ -1301,6 +1301,10 @@ class TypedColumnWriterImpl : public ColumnWriterImpl, public TypedColumnWriter<
     bool single_nullable_element =
         (level_info_.def_level == level_info_.repeated_ancestor_def_level + 1) &&
         leaf_field_nullable;
+    if (!leaf_field_nullable && leaf_array.null_count() != 0) {
```
Comment (wgtmac) [inline cpp/src/parquet/column_writer.cc:1304]: Should we take special care when it is `kUnknownNullCount`? It is possible for an end user to produce an array without setting a precise null_count value.

Comment (wgtmac) [reply cpp/src/parquet/column_writer.cc:1304]: I just found that `Array::Validate()` and `Array::ValidateFull()` are checking `DataType` not `Field`. If it knows the latter, we can also validate null_count with nullable from that Field.

**Review by wgtmac — APPROVED** [review]
The fix looks good. I have some minor comments.

File: cpp/src/parquet/column_writer.cc (hunk 1 above)
Comment (mapleFU) [reply cpp/src/parquet/column_writer.cc:1304]: > Should we take special care when it is kUnknownNullCount? It is possible for an end user to produce an array without setting a precise null_count value.

`kUnknownNullCount` would be counted when it's that value

Comment (wgtmac) [reply cpp/src/parquet/column_writer.cc:1304]: Yes in most cases. I remember that when I did the trick to make mutable (and reusable) arrays. An empty ArrayData is first created by calling `std::shared_ptr<ArrayData> ArrayData::Make(std::shared_ptr<DataType> type, int64_t length, int64_t null_count, int64_t offset)` with default values and then set it to array via `void Array::SetData(const std::shared_ptr<ArrayData>& data)`. The null_count is -1 if not manually maintained.

Comment (mapleFU) [reply cpp/src/parquet/column_writer.cc:1304]:
```c++
int64_t ArrayData::GetNullCount() const {
  int64_t precomputed = this->null_count.load();
  if (ARROW_PREDICT_FALSE(precomputed == kUnknownNullCount)) {
    if (this->buffers[0]) {
      precomputed = this->length -
                    CountSetBits(this->buffers[0]->data(), this->offset, this->length);
    } else {
      precomputed = 0;
    }
    this->null_count.store(precomputed);
  }
  return precomputed;
}
```

`null_count()` call will generate null_count if not exists.

Comment (mapleFU) [inline cpp/src/parquet/column_writer.cc:1304]: I don't know should we check `single_nullable_element` rather than `leaf_field_nullable`

Comment (pitrou) [reply cpp/src/parquet/column_writer.cc:1304]: > I just found that `Array::Validate()` and `Array::ValidateFull()` are checking `DataType` not `Field`. If it knows the latter, we can also validate null_count with nullable from that Field.

See https://github.com/apache/arrow/issues/31387

**Comment by pitrou** [comment]
Java JNI failure looks related.

File: cpp/src/parquet/column_writer.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/column_writer.cc:1304]: I don't even know what `single_nullable_element` means.

**Comment by lidavidm** [comment]
The Java code has been moved, so we could ignore the JNI failure here and open an issue in the other repo (I'm still getting all the CI set up, though)

**Comment by pitrou** [comment]
Well, I think I got the failure fixed anyway (at least locally it works, let's wait for CI).

File: cpp/src/parquet/column_writer.cc (hunk 1 above)
Comment (mapleFU) [reply cpp/src/parquet/column_writer.cc:1304]: As what parquet handle nulls, `leaf_field_nullable` might including nulls in parents?

**Comment by pitrou** [comment]
@lidavidm @jduo Are the Java changes ok?

File: cpp/src/parquet/column_writer.cc (hunk 1 above)
Comment (mapleFU) [reply cpp/src/parquet/column_writer.cc:1304]: E.g. `List< int notnull> nullable`, we may have the `single_nullable_element == false` but `leaf_field_nullable == true`?

**Review by lidavidm — APPROVED** [review]

File: cpp/src/parquet/column_writer.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/column_writer.cc:1304]: `leaf_field_nullable` is computed from `PathBuilder.nullable_in_parent_`, which itself is initialized in e.g.:
https://github.com/apache/arrow/blob/fb8e8122f623f4548b22ece7485c4570d7ece1a6/cpp/src/parquet/arrow/path_internal.cc#L792-L801

Comment (pitrou) [reply cpp/src/parquet/column_writer.cc:1304]: And for lists:
https://github.com/apache/arrow/blob/fb8e8122f623f4548b22ece7485c4570d7ece1a6/cpp/src/parquet/arrow/path_internal.cc#L725-L741

**Pull request merged**